src/solver/optim/adam.cpp

```diff
@@ -0,0 +1,218 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2023 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/optim/adam/solvers.hpp>
+
+#include <miopen/adam.hpp>
+#include <miopen/datatype.hpp>
+#include <miopen/kernel_build_params.hpp>
+#include <miopen/optim/adam/invoke_params.hpp>
+#include <miopen/target_properties.hpp>
+
+namespace miopen {
+
+namespace solver {
+
+namespace adam {
+
+bool Adam::IsApplicable([[maybe_unused]] const ExecutionContext& context,
+                        const miopen::adam::ProblemDescription& problem) const
+{
+    if(!problem.IsAllPacked())
+        return false;
+    return true;
+}
+
+inline size_t AlignUp(size_t num, size_t align) { return (num + align - 1) / align * align; }
+
+ConvSolution Adam::GetSolution([[maybe_unused]] const ExecutionContext& context,
+                               const miopen::adam::ProblemDescription& problem) const
+{
+    auto result = ConvSolution{miopenStatusSuccess};
+
+    {
+        auto param_dtype = miopen::GetDataType(problem.GetParamDesc().GetType());
+        auto grad_dtype  = miopen::GetDataType(problem.GetGradDesc().GetType());
```

Comment:
If `IsAmp() == false`, do we have GradDesc? AFAICS it is unused (or its type is doesn't matter, I am not aware of all the details).
```suggestion
        auto grad_dtype  = problem.IsAmp() //
            ? miopen::GetDataType(problem.GetGradDesc().GetType()) //
            : float; // unused dummy
```

Reply:
@et16kr The idea of this comment is to make code more clear for the reader. Am I right that `grad_dtype` is unused when `problem.IsAmp()` is true? If so, then initialization of `grad_dtype` with value of `param_dtype` does not help much to figure that out.

Reply:
[not resolved yet]

Reply:
When step tensor is used, the amp adam kernel is invoked even if ```IsAmp() != true```. 
Therefore, a condition for step tensor has been added.